src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,562 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
         lvaSetVarAddrExposed(lclNum DEBUGARG(AddressExposedReason::ESCAPE_ADDRESS));
     }
 }
+
+#if defined(TARGET_ARM64)
+
+//-----------------------------------------------------------------------------
+// UpdateStoreWeight: Updates the weighting to take account of a local store.
+//
+// Arguments:
+//     hasConvertFromMask - Is this the store of a convert from mask
+//     blockWeight - Weight of the block the store is contained in
+//
+void Compiler::LCLMasksWeight::UpdateStoreWeight(bool hasConvertFromMask, weight_t blockWeight)
+{
+    if (hasConvertFromMask)
+    {
+        // Count the cost of the existing convert mask to vector.
+        weight_t incVal = blockWeight * costOfConvertMaskToVector;
+        JITDUMP("Incrementing currentCost by %f. ", incVal);
+        currentCost += incVal;
+    }
+    else
+    {
+        // Switching would require adding a convert vector to mask.
+        weight_t incVal = blockWeight * costOfConvertVectorToMask;
+        JITDUMP("Incrementing switchCost by %f. ", incVal);
+        switchCost += incVal;
+    }
+    DumpTotalWeight();
+}
+
+//-----------------------------------------------------------------------------
+// UpdateVarWeight: Updates the weighting to take account of a local variable use.
+//
+// Arguments:
+//     hasConvertFromMask - Is this variable converted to a mask when used
+//     blockWeight - Weight of the block the use is contained in
+//
+void Compiler::LCLMasksWeight::UpdateVarWeight(bool hasConvertToMask, weight_t blockWeight)
+{
+    if (hasConvertToMask)
+    {
+        // Count the cost of the existing convert vector to mask.
+        weight_t incVal = blockWeight * costOfConvertVectorToMask;
+        JITDUMP("Incrementing currentCost by %f. ", incVal);
+        currentCost += incVal;
+    }
+    else
+    {
+        // Switching would require adding a convert mask to vector.
+        weight_t incVal = blockWeight * costOfConvertMaskToVector;
+        JITDUMP("Incrementing switchCost by %f. ", incVal);
+        switchCost += incVal;
+    }
+    DumpTotalWeight();
+}
+
+//-----------------------------------------------------------------------------
+// CacheSimdTypes: Cache the simd types of a hwintrinsic
+//
+// Arguments:
+//     op - The HW intrinsic to cache
+//
+void Compiler::LCLMasksWeight::CacheSimdTypes(GenTreeHWIntrinsic* op)
+{
+    CorInfoType newSimdBaseJitType = op->GetSimdBaseJitType();
+    unsigned    newSimdSize        = op->GetSimdSize();
+
+    assert((newSimdBaseJitType != CORINFO_TYPE_UNDEF));
+    assert((simdBaseJitType == CORINFO_TYPE_UNDEF) ||
+           ((simdBaseJitType == newSimdBaseJitType) && (simdSize == newSimdSize)));
```

Comment:
I don't think this can be an assert. This does not necessarily have to match between separate stores to the local. As an example you might try something like
```csharp
Vector<int> v = Environment.TickCount % 2 switch {
  0 => Sve.CreateTrueMaskInt32(),
  1 => Unsafe.BitCast<Vector<uint>, Vector<int>>(Sve.CreateTrueMaskUInt32()),
};
```
(or even a mask of a different size).

Reply:
I don't really understand why `Sve.CreateTrueMaskUInt32()` is typed as `Vector<uint>`. Isn't it basically just a bitmask with the first N bits set? Does it make any sense to treat it as a vector with `uint` elements?

Reply:
> Isn't it basically just a bitmask with the first N bits set? 

Not quite. When in a predicate register it's a vector of N bits, where N is the number of `uint`s in a `vector<uint>`. When it's moved to a vector then each bit is moved to a different lane in the `vector<uint>`.

When moving a vector back to a mask, each lane in the vector is `cmpne` to `0`.


```
vector<uint> m = Sve.CreateTrueMaskUInt32(); // mask <1b,1b,1b,1b...> converted to vector [1,1,1,1...]
vector<uint> v1 = Vector.Create<uint>(5); // [5,5,5,5...]
vector<uint> v2 = Sve.Add(m, v1); // [6,6,6,6...]
Sve.SomeMethodThatUsesAMask(m); //  [6,6,6,6...] converted to mask <1b,1b,1b,1b...>
```